auth/api/iam/api_mitzxnuts.go

```diff
@@ -0,0 +1,464 @@
+package iam
+
+import (
+	"context"
+	crypt "crypto"
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/auth/log"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/crypto"
+	http2 "github.com/nuts-foundation/nuts-node/http"
+	"github.com/nuts-foundation/nuts-node/storage"
+	"github.com/nuts-foundation/nuts-node/vcr/holder"
+	"github.com/nuts-foundation/nuts-node/vcr/pe"
+	"github.com/nuts-foundation/nuts-node/vcr/verifier"
+	"github.com/nuts-foundation/nuts-node/vdr/didweb"
+	"github.com/nuts-foundation/nuts-node/vdr/resolver"
+	"io"
+	"maps"
+	"net/http"
+	"net/url"
+	"slices"
+	"time"
+)
+
+const oid4vicSessionValidity = 15 * time.Minute
+
+type Oid4vciSession struct {
+	HolderDid   string
+	IssuerDid   string
+	RedirectUrl string
+	RedirectUri string
+}
+
+func (r Wrapper) StartOid4vciIssuance(ctx context.Context, request StartOid4vciIssuanceRequestObject) (StartOid4vciIssuanceResponseObject, error) {
+	if request.Body == nil {
+		// why did oapi-codegen generate a pointer for the body??
+		return nil, core.InvalidInputError("missing request body")
+	}
+	requestHolder, err := did.ParseDID(request.Did)
+	if err != nil {
+		return nil, core.NotFoundError("did not found: %w", err)
+	}
+	isWallet, err := r.vdr.IsOwner(ctx, *requestHolder)
+	if err != nil {
+		return nil, err
+	}
+	if !isWallet {
+		return nil, core.InvalidInputError("did not owned by this node: %w", err)
+	}
+
+	issuerDid, err := did.ParseDID(request.Body.Issuer)
+	if err != nil {
+		return nil, core.NotFoundError("did not found: %w", err)
+	}
+
+	metadata, err := r.auth.MitzXNutsIAMClient().OpenIdCredentialIssuerMetadata(ctx, *issuerDid)
+	if err != nil {
+		return nil, err
+	}
+	if len(metadata.AuthorizationServers) == 0 {
+		return nil, core.NotFoundError("cannot locate any authorization endpoint in %s", issuerDid.String())
+	}
+	for i := range metadata.AuthorizationServers {
+		// TODO: do some kind of logic here on supported credentials
+		serverURL, err := url.Parse(metadata.AuthorizationServers[i])
+		if err != nil {
+			return nil, err
+		}
+		metadataFromUrl, err := r.auth.MitzXNutsIAMClient().OpenIdConfiguration(ctx, *serverURL)
```

Comment:
shouldn't this be the authorization server metadata instead of openid?

Reply:
Nope, it is the OpenIdConfiguration of the _issuer_ authorization server. To my understanding it works like this:

 * The issuer publishes a file at `.well-known/openid-credential-issuer`
 * This file contains a list of authorization servers, these are plain old oid(c) authorization servers
 * The first of the list is picked and queried for `/.well-known/openid-configuration`
 * The /authorization endpoint of that server is used to authorize.

This is how I understood the oidc4vci documentation:

> (2) The Wallet uses the Credential Issuer's URL to fetch the Credential Issuer metadata, as described in [Section 11.2](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#credential-issuer-metadata). The Wallet needs the metadata to learn the Credential types and formats that the Credential Issuer supports and to determine the Authorization Endpoint (OAuth 2.0 Authorization Server) as well as Credential Endpoint required to start the request. This specification enables deployments where the Credential Endpoint and the Authorization Endpoint are provided by different entities. Please note that in this example, the Credential Issuer and OAuth 2.0 Authorization Server correspond to the same entity.

Reply:
so they chose for openid-configuration for issuance and auth server metadata for presentation. Annoying, but ok. They might come together in the future.